core/rawdb/accessors_indexes.go

```diff
@@ -139,15 +140,41 @@ func ReadTransaction(db ethdb.Reader, hash common.Hash) (*types.Transaction, com
 	if blockHash == (common.Hash{}) {
 		return nil, common.Hash{}, 0, 0
 	}
-	body := ReadBody(db, blockHash, *blockNumber)
-	if body == nil {
+	bodyRLP := ReadBodyRLP(db, blockHash, *blockNumber)
+	if bodyRLP == nil {
 		log.Error("Transaction referenced missing", "number", *blockNumber, "hash", blockHash)
 		return nil, common.Hash{}, 0, 0
 	}
-	for txIndex, tx := range body.Transactions {
-		if tx.Hash() == hash {
-			return tx, blockHash, *blockNumber, uint64(txIndex)
+	txnListRLP, _, err := rlp.SplitList(bodyRLP)
+	if err != nil {
+		return nil, common.Hash{}, 0, 0
+	}
+	txnIterator, err := rlp.NewListIterator(txnListRLP)
+	if err != nil {
+		return nil, common.Hash{}, 0, 0
+	}
+
+	txIndex := 0
```

Comment:
Changes in this function should be submitted in their own PR. It's kind of an independent optimization.